Cabal-described/src/Distribution/Described/Extension.hs

```diff
@@ -0,0 +1,508 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+module Distribution.Described.Extension
+  ( reEnableExtension
+  , reKnownExtension
+  , reDisableExtension
+  , reXs
+      -- * Extension groups
+  , xGroupInteractive
+  , xGroupPhase
+  , xGroupSyntax
+  , xGroupImportExport
+  , xGroupTypes
+  , xGroupRecords
+  , xGroupDeriving
+  , xGroupPatterns
+  , xGroupClassesInstances
+  , xGroupLiterals
+  , xGroupConstraints
+  , xGroupTypeSignatures
+  , xGroupBindingsGeneralisation
+  , xGroupTemplates
+  , xGroupBangStrict
+  , xGroupParallelConcurrent
+  , xGroupUnboxedPrimitive
+  , xGroupForeign
+  , xGroupSafe
+  , xGroupMiscellaneous
+  , xGroupBugs
+  , xUngrouped
+  ) where
+
+import Data.List ((\\))
+import Data.String (IsString (..))
+import Distribution.Pretty (prettyShow)
+import Distribution.Described
+import Language.Haskell.Extension
+
+instance Described Extension where
+    describe _ = REUnion
+        [ RENamed "enable-extension" reKnownExtension
+        , RENamed "disable-extension" reDisableExtension
```

Comment:
I repeat myself: **this is wrong**. The grammar accapts any tokens. The extensions specified may be anything, something which particular `Cabal` version doesn;t know about. The list of extensions **is not part of .cabal file specification**.

Listing all known extensions in the grammar specification is a bad idea.

Reply:
How about something like this with its "unknown-extension"?

```haskell
instance Described Extension where
    describe _ = REUnion
        [ reEnableExtension
        , reDisableExtension
        , reUnknownExtension
        ]

reEnableExtension :: GrammarRegex a
reEnableExtension = RENamed "enable-known-extension" reKnownExtension

reDisableExtension :: GrammarRegex a
reDisableExtension = REUnion ["No" <> reEnableExtension]

reUnknownExtension :: GrammarRegex a
reUnknownExtension = RENamed "unknown-extension" RETodo
```

Reply:
No. Grammar doesn't care about extension names. They don't belong there.

The `KnownExtension` is an optimization: a constructor tag (machine integer) is better than a `String` (or `Text`). Perfectly, `Cabal` won't even have those exposed at all (for example it would allow smoother GHC-updates, as adjusting `KnownExtension` currently is a breaking change, but it would be better if it wasnt).

If people want extension list, add a link to the GHC manual.